utils/keywait.1
.\" keywait - pause and wait for keyboard input
.\" Copyright (C) 2013-2014 David Sugar <dyfet@gnutelephony.org>
.\"
.\" This manual page is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 3 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU Lesser General Public License
.\" along with this program.  If not, see <http://www.gnu.org/licenses/>.
.\"
.\" This manual page is written especially for Debian GNU/Linux.
.\"
.TH keywait "1" "August 2013" "GNU uCommon" "GNU Telephony"
.SH NAME
keywait \- pause and wait for input.
.SH SYNOPSIS
.B keywait
.RI [ options ]
.RI [ prompt ]
.br
.SH DESCRIPTION
Pause and wait for keyboard input.  A prompt can be displayed to the user, 
and the timeout to wait for keyboard input can be specified.  Before waiting
any pending input is automatically flushed.  If no timeout is specified then
keywait will wait forever.
.SH OPTIONS
.TP
.BI \-\-timeout= seconds
Specify number of seconds to wait for input.
.TP
.B \-\-help
Outputs help screen for the user.
.SH EXIT CODES
Returns 0 if keyboard entry was received.  1 if timeout.  2 for any error.
.SH AUTHOR
.B keywait
was written by David Sugar <dyfet@gnutelephony.org>.
.SH "REPORTING BUGS"
Report bugs to bug-commoncpp@gnu.org or bugs@gnutelephony.org.
.SH COPYRIGHT
Copyright \(co 2013-2014 David Sugar, Tycho Softworks.
.br
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR
PURPOSE.



NEWS
uCommon Library NEWS -- history of visible user changes. Updated 2008-04-15.

uCommon C++ 0.4 codebase completed and ready for release with the introduction
of the GNU GPL V3.  This is a complete rewrite of UCommon as a C++ library for
threading and design patterns suitable for deeply embedded targets.

uCommon C++ 0.7 codebase now has core documentation and basic library theory in
header files.  This documentation can be generated with doxygen or by using
"make doxy".

uCommon C++ 1.0 release has been simplified, fully documented, and modified in
specific ways to simplify the migration and merge with GNU Common C++ to form
GNU Common C++ 2.0.

uCommon C++ 1.5 release begins the process of migrating GNU Common C++ into the
core GNU uCommon C++ codebase. Some GNU Common C++ classes, including tcp
streaming, xml parser, and persist, have been revised and integrated for the
1.5.0 release itself.  Other GNU Common C++ classes will be integrated as
requested and needed.  The 1.5/1.x series will actually be the last UCommon
release sets, as GNU uCommon C++ 2.0 will instead be released as GNU Common C++
2 2.0.

uCommon C++ 2.0 released under the GNU Lesser General Public License, Version 3
or later.

uCommon C++ 2.1 is considered a refactored abi release offering greater
clarity, consistency of use, and additional features that were migrated from
Common C++. The 2.1 release series may continue to introduce additional
classes.

ucommon 2.2 merges other ucommon related libraries to create a single ucommon
framework for application development.



utils/keywait.cpp
// Copyright (C) 2010-2014 David Sugar, Tycho Softworks.
//
// This file is part of GNU uCommon C++.
//
// GNU uCommon C++ is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published
// by the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// GNU uCommon C++ is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with GNU uCommon C++.  If not, see <http://www.gnu.org/licenses/>.

#include <ucommon/secure.h>

#ifdef  _MSWINDOWS_
#include <conio.h>
#else
#include <unistd.h>
#include <termios.h>
#endif

using namespace UCOMMON_NAMESPACE;

static shell::flagopt helpflag('h',"--help",    _TEXT("display this list"));
static shell::flagopt althelp('?', NULL, NULL);
static shell::numericopt timeout('t', "--timeout", _TEXT("optional keyboard input timeout"), "seconds", 0);

#ifdef _MSWINDOWS_

HANDLE input;
DWORD mode;

#else

static struct termios orig, current;

static void cleanup(void)
{
    tcsetattr(0, TCSANOW, &orig);
}
#endif

static void keyflush(void)
{
#ifdef  _MSWINDOWS_
    TCHAR ch;
    DWORD count;

    while(WaitForSingleObject(input, 20) == WAIT_OBJECT_0) {
        ReadConsole(input, &ch, 1, &count, NULL);
    }
#else
    fd_set inp;
    struct timeval tv;
    char ch;

    FD_ZERO(&inp);
    FD_SET(0, &inp);
    
    for(;;) {
        tv.tv_usec = 20000;
        tv.tv_sec = 0;
        if(select(1, &inp, NULL, NULL, &tv) <= 0)
            break;
        if(::read(0, &ch, 1) < 1)
            break;
    }
#endif
}

PROGRAM_MAIN(argc, argv)
{
    shell::bind("keywait");
    shell args(argc, argv);
    bool nl = false;

    unsigned count = 0;
    int excode = 2;

    if(is(helpflag) || is(althelp)) {
        printf("%s\n", _TEXT("Usage: keywait [options] [text]"));
        printf("%s\n\n", _TEXT("Wait for keyboard input"));
        printf("%s\n", _TEXT("Options:"));
        shell::help();
        printf("\n%s\n", _TEXT("Report bugs to dyfet@gnutelephony.org"));
        PROGRAM_EXIT(0);
    }

    while(count < args()) {
        shell::printf("%s ", args[count++]);
        nl = true;
    }

#ifdef  _MSWINDOWS_
    input = GetStdHandle(STD_INPUT_HANDLE);
    GetConsoleMode(input, &mode);
    SetConsoleMode(input, 0);

    keyflush();

    DWORD timer = INFINITE;

    if(is(timeout))
        timer = *timeout;

    if(WaitForSingleObject(input, timer) == WAIT_OBJECT_0) {
        keyflush();
        excode = 0;
    }
    else
        excode = 1;

    SetConsoleMode(input, mode);

#else
    tcgetattr(0, &orig);
    shell::exiting(&cleanup);

    tcgetattr(0, &current);
    cfmakeraw(&current);
    tcsetattr(0, TCSANOW, &current);
    fd_set inp;
    struct timeval tv = {0, 0};
    struct timeval *tvp = NULL;

    FD_ZERO(&inp);
    FD_SET(0, &inp);
    

    if(*timeout) {
        tv.tv_sec = *timeout;
        tvp = &tv;
    }

    int rtn = select(1, &inp, NULL, NULL, tvp);
    if(rtn > 0) {
        keyflush();
        excode = 0;
    }
    else {
        if(rtn == 0)
            excode = 1;
    }
    cleanup();
#endif
    if(nl)
        shell::printf("\n");

    PROGRAM_EXIT(excode);
}

